File: apps/api/app/api/v1/assets.py
"""Media upload, confirmation, download and listing.

The upload architecture is the point of this module (directive §13):

    Browser ──presigned PUT──▶ Object storage

Bytes never pass through Next.js or FastAPI. The API issues a short-lived signed
URL and records a `pending` row; the browser uploads directly; a confirm call
verifies what landed and marks the asset usable.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Query, Response, status

from app.api.deps import CurrentUser, DbSession, Pagination, RedisClient
from app.core.config import settings
from app.core.enums import AssetKind, AssetSource, AssetStatus
from app.core.errors import ValidationFailed
from app.core.logging import get_logger
from app.schemas.asset import (
    AssetConfirmRequest,
    AssetPublic,
    DownloadUrlResponse,
    MediaCounts,
    PresignedUploadResponse,
    UploadUrlRequest,
    UploadUrlResponse,
)
from app.schemas.common import Page
from app.services import rate_limit
from app.services.storage import AssetService

logger = get_logger(__name__)
router = APIRouter(tags=["media"])

_UPLOAD_RATE_LIMIT = 60
_UPLOAD_RATE_WINDOW = 60


@router.post(
    "/assets/upload-url",
    response_model=UploadUrlResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Get a direct-upload URL",
)
async def create_upload_url(
    payload: UploadUrlRequest,
    session: DbSession,
    redis: RedisClient,
    user: CurrentUser,
    response: Response,
) -> UploadUrlResponse:
    """Validates the declared type and size, then issues a presigned PUT.

    Rate-limited because issuing URLs is cheap for us and expensive to abuse:
    each one is a writable handle into the bucket for its lifetime.
    """
    await rate_limit.check_request_rate(
        redis,
        user.id,
        action="upload_url",
        limit=_UPLOAD_RATE_LIMIT,
        window_seconds=_UPLOAD_RATE_WINDOW,
    )

    service = AssetService(session)
    asset, presigned = await service.create_upload(
        user_id=user.id,
        filename=payload.filename,
        content_type=payload.content_type,
        kind=payload.kind,
        size_bytes=payload.size_bytes,
    )
    await session.commit()

    response.headers["Cache-Control"] = "no-store"
    return UploadUrlResponse(
        asset_id=asset.id,
        upload=PresignedUploadResponse(
            url=presigned.url,
            method=presigned.method,
            headers=presigned.headers,
            expires_in=presigned.expires_in,
        ),
        confirm_url=f"{settings.api_v1_prefix}/assets/{asset.id}/confirm",
    )


@router.post(
    "/assets/{asset_id}/confirm", response_model=AssetPublic, summary="Confirm an upload"
)
async def confirm_upload(
    asset_id: uuid.UUID,
    payload: AssetConfirmRequest,
    session: DbSession,
    user: CurrentUser,
    response: Response,
) -> AssetPublic:
    """Verifies the object landed and matches its declared size, then marks it
    ready. Until this succeeds the asset cannot be used as a generation input."""
    service = AssetService(session)
    asset = await service.require_for_user(asset_id, user.id)
    confirmed = await service.confirm_upload(
        asset=asset,
        duration_seconds=payload.duration_seconds,
        width=payload.width,
        height=payload.height,
    )
    await session.commit()
    response.headers["Cache-Control"] = "no-store"
    return service.to_public(confirmed)


@router.post(
    "/assets/{asset_id}/download-url",
    response_model=DownloadUrlResponse,
    summary="Get a download URL",
)
async def create_download_url(
    asset_id: uuid.UUID,
    session: DbSession,
    user: CurrentUser,
    response: Response,
) -> DownloadUrlResponse:
    """Issues a short-lived signed GET.

    A signed URL, not a public object: the bucket stays private, so a leaked
    link expires instead of exposing a user's media indefinitely.
    """
    service = AssetService(session)
    asset = await service.require_for_user(asset_id, user.id)

    if asset.status != AssetStatus.READY:
        raise ValidationFailed("That media item is not ready yet.")

    response.headers["Cache-Control"] = "no-store"
    return DownloadUrlResponse(
        url=service.download_url(asset, as_attachment=True),
        expires_in=settings.storage_presign_expiry_seconds,
    )


@router.get("/media", response_model=Page[AssetPublic], summary="Media library")
async def list_media(
    session: DbSession,
    user: CurrentUser,
    page: Pagination,
    response: Response,
    kind: Annotated[AssetKind | None, Query()] = None,
    source: Annotated[AssetSource | None, Query()] = None,
) -> Page[AssetPublic]:
    service = AssetService(session)
    assets, next_cursor, has_more = await service.repo.list_for_user(
        user.id, page=page, kind=kind, source=source
    )
    response.headers["Cache-Control"] = "no-store"
    return Page[AssetPublic](
        items=[service.to_public(asset) for asset in assets],
        next_cursor=next_cursor,
        has_more=has_more,
    )


@router.get("/media/counts", response_model=MediaCounts, summary="Media tab counts")
async def media_counts(session: DbSession, user: CurrentUser) -> MediaCounts:
    """Counts per kind, in one grouped query rather than four requests."""
    service = AssetService(session)
    counts = await service.repo.count_by_kind(user.id)
    return MediaCounts(
        all=sum(counts.values()),
        video=counts.get("video", 0),
        image=counts.get("image", 0),
        audio=counts.get("audio", 0),
    )

File: apps/web/src/services/assets.ts
import {
  assetSchema,
  mediaCountsSchema,
  pageSchema,
  uploadUrlSchema,
  type Asset,
  type AssetKind,
  type MediaCounts,
  type Page,
} from "@zolexai/workflow-contracts";
import { apiFetch, uploadToStorage } from "@/lib/api/client";
import { z } from "zod";

const assetPageSchema = pageSchema(assetSchema);
const downloadUrlSchema = z.object({ url: z.string(), expires_in: z.number().int() });

export interface ListMediaParams {
  limit?: number;
  cursor?: string | null;
  kind?: AssetKind | null;
  source?: "upload" | "generated" | null;
}

export async function listMedia(
  params: ListMediaParams = {},
  signal?: AbortSignal,
): Promise<Page<Asset>> {
  const query = new URLSearchParams();
  query.set("limit", String(params.limit ?? 24));
  if (params.cursor) query.set("cursor", params.cursor);
  if (params.kind) query.set("kind", params.kind);
  if (params.source) query.set("source", params.source);
  return apiFetch(`/media?${query.toString()}`, assetPageSchema, { signal });
}

export async function fetchMediaCounts(signal?: AbortSignal): Promise<MediaCounts> {
  return apiFetch("/media/counts", mediaCountsSchema, { signal });
}

/** One asset the user owns — used when a screen has an id but not the asset. */
export async function fetchAsset(assetId: string, signal?: AbortSignal): Promise<Asset> {
  return apiFetch(`/assets/${assetId}`, assetSchema, { signal });
}

export async function fetchDownloadUrl(assetId: string): Promise<string> {
  const data = await apiFetch(`/assets/${assetId}/download-url`, downloadUrlSchema, {
    method: "POST",
  });
  return data.url;
}

/**
 * Uploads a file directly to object storage.
 *
 *   1. ask the API for a presigned PUT (it validates type and size first)
 *   2. PUT the file straight to storage
 *   3. confirm, so the API can verify what actually landed
 *
 * Step 2 is the important one: the bytes never pass through Next.js or FastAPI
 * (directive §13). A 500 MB video routed through the API would occupy a server
 * process for the whole transfer.
 *
 * Until step 3 succeeds the asset stays `pending` and cannot be used as a
 * generation input — so a half-finished upload can never reach a worker.
 */
export async function uploadAsset(
  file: File,
  kind: AssetKind,
  options: { signal?: AbortSignal } = {},
): Promise<Asset> {
  const ticket = await apiFetch("/assets/upload-url", uploadUrlSchema, {
    method: "POST",
    signal: options.signal,
    body: {
      filename: file.name,
      content_type: file.type || "application/octet-stream",
      kind,
      size_bytes: file.size,
    },
  });

  await uploadToStorage(
    ticket.upload.url,
    ticket.upload.method,
    ticket.upload.headers,
    file,
    options.signal,
  );

  return apiFetch(`/assets/${ticket.asset_id}/confirm`, assetSchema, {
    method: "POST",
    signal: options.signal,
    body: {},
  });
}

/** Maps a browser MIME type onto the asset kind the API expects. */
export function kindForFile(file: File): AssetKind | null {
  if (file.type.startsWith("video/")) return "video";
  if (file.type.startsWith("image/")) return "image";
  if (file.type.startsWith("audio/")) return "audio";
  return null;
}

export function formatBytes(bytes: number | null): string {
  if (!bytes) return "—";
  const units = ["B", "KB", "MB", "GB"];
  let value = bytes;
  let unit = 0;
  while (value >= 1024 && unit < units.length - 1) {
    value /= 1024;
    unit += 1;
  }
  return `${value.toFixed(value >= 10 || unit === 0 ? 0 : 1)} ${units[unit]}`;
}

export function formatDuration(seconds: number | null): string | null {
  if (seconds == null) return null;
  const total = Math.round(seconds);
  return `${Math.floor(total / 60)}:${String(total % 60).padStart(2, "0")}`;
}
